gnomad_qc/v4/annotations/generate_freq.py

```diff
@@ -0,0 +1,441 @@
+
+    :param mt: Hail Table containing freq, ab_adjusted_freq, high_ab_het annotations.
+    :return: Hail MatirxTable with faf & popmax annotations.
+    """
+    faf, faf_meta = faf_expr(
+        mt.ab_adjusted_freq, mt.freq_meta, mt.locus, POPS_TO_REMOVE_FOR_POPMAX
+    )
+    mt = mt.annotate_rows(
+        faf=faf,
+        popmax=pop_max_expr(
+            mt.ab_adjusted_freq, mt.freq_meta, POPS_TO_REMOVE_FOR_POPMAX
+        ),
+    )
+    mt = mt.annotate_globals(
+        faf_meta=faf_meta,
+        faf_index_dict=make_faf_index_dict(faf_meta, label_delimiter="-"),
+    )
+    mt = mt.annotate_rows(
+        popmax=mt.popmax.annotate(
+            faf95=mt.faf[
+                mt.faf_meta.index(lambda x: x.values() == ["adj", mt.popmax.pop])
+            ].faf95
+        )
+    )
+    return mt
+
+
+def main(args):  # noqa: D103
+    # TODO: Determine if splitting subset freq from whole callset agg
+    subsets = args.subsets
+    test = args.test
+    chrom = args.chrom
+    af_threshold = args.af_threshold
+    adjust_freqs = args.adjust_freqs
+
+    hl.init(
+        log=f"/generate_frequency_data{'.' + '_'.join(subsets) if subsets else ''}.log",
+        default_reference="GRCh38",
+        tmp_dir="gs://gnomad-tmp-4day",
+    )
+    # TODO: Update to release = True once outlier filtering is complete,
+    # possibly sample_meta=True if added
+    vds = get_gnomad_v4_vds(test=test)
+    meta_ht = meta.ht()
+    final_anns = []
+
+    logger.info("Adding metadata to VDS variant data cols...")
+    vds = hl.vds.VariantDataset(
+        vds.reference_data,
+        vds.variant_data.annotate_cols(meta=meta_ht[vds.variant_data.col_key]),
+    )
+
+    if test or chrom:
+        if test:
+            logger.info("Filtering to DRD2 in test VDS for testing purposes...")
+            test_interval = [
+                hl.parse_locus_interval("chr11:113409605-113475691")
+            ]  # NOTE: test on gene DRD2 for now, will update to chr22 when original PR goes in
+            vds = hl.vds.filter_intervals(vds, test_interval)
+            variants, samples = vds.variant_data.count()
+            logger.info(
+                "Test VDS has %s variants in DRD2 in %s samples...", variants, samples
+            )
+        else:
+            logger.info("Filtering to chromosome %s...")
+            vds = hl.vds.filter_chromosomes(vds, keep=f"chr{chrom}")
+
+    logger.info("Annotating non_ref hets pre-split...")
+    vds = annotate_non_ref_het(vds)
+
+    # if args.subsets:
+    #     vds = hl.vds.filter_samples(
+    #         vds,
+    #     )  # TODO: Where is subset information coming meta.project_meta.ukb is defined for non-ukb and ukd, how do we determine non-topmed v3.1 meta did (topmed=ht.s.startswith("NWD"))
+
+    logger.info("Splitting VDS....")
+    vds = hl.vds.split_multi(vds, filter_changed_loci=True)
+
+    logger.info("Densifying VDS...")
+    mt = hl.vds.to_dense_mt(vds)
```

Comment:
Again, not for this run, but could also add this as an option to the `get_gnomad_v4_vds`